libraries/lib-graphics/GraphicsDataCache.h

```diff
@@ -0,0 +1,327 @@
+   {
+      GraphicsDataCacheKey Key;
+      GraphicsDataCacheElementBase* Data {};
+   };
+   //! Cache lookup is a vector, with items sorted using Key
+   using Lookup = std::vector<LookupElement>;
+
+   //! A result of the cache lookup
+   struct GRAPHICS_API BaseLookupResult final
+   {
+      //! Iterator to the first cache element that fulfills the request
+      Lookup::iterator Begin;
+      //! Iterator past the last cache element that fulfills the request
+      Lookup::iterator End;
+      //! First column of the first element that matches the request
+      size_t LeftOffset {};
+      //! Offset from the right for the last element that matches the request
+      size_t RightOffset {};
+   };
+   //! Create a new Cache element. Implementation is responsible of the lifetime control.
+   virtual GraphicsDataCacheElementBase* CreateElement(const GraphicsDataCacheKey& key) noexcept = 0;
+   //! This method is called, when the cache element should be evicted. Implementation may not deallocate the object.
+   virtual void DisposeElement(GraphicsDataCacheElementBase* element) = 0;
+
+   //! This method is called on all elements matching the request that are not complete (i. e. IsComplete if false).
+   virtual bool UpdateElement(
+      const GraphicsDataCacheKey& key, GraphicsDataCacheElementBase& element) = 0;
+
+   //! Perform a lookup inside the cache. This method modifies mLookup and invalidates any previous result.
+   BaseLookupResult PerformBaseLookup(const ZoomInfo& zoomInfo, double t0, double t1);
+private:
+   // Called internally to create a list of items in the mNewLookupItems
+   bool CreateNewItems();
+   // Called internally if the CreateNewItems has failed to dispose all the elements created
+   void DisposeNewItems();
+   Lookup::iterator FindKey(GraphicsDataCacheKey key);
+
+   // Called internally to evict the no longer needed items. Cache keeps mCacheSizeMultiplier * mMaxWidth / CacheElementWidth items
+   void PerformCleanup();
+   // A heap based approach if cache needs to evict more than one item
+   void PerformFullCleanup(int64_t currentSize, int64_t itemsToEvict);
+   // This vector is sorted according to the key
+   Lookup mLookup;
+   // This vector is used to avoid memory allocations when growing the cache
+   Lookup mLookupHelper;
+   // This is a helper vector to store newly created items before they are inserted in cache
+   Lookup mNewLookupItems;
+   // This is a helper vector to implement the heap structure for the LRU policy
+   std::vector<size_t> mLRUHelper;
+
+   // Sample rate associated with this cache
+   double mSampleRate {}; // DV: Why do we use double for sample rate? I don't know
+
+   // The max width of the request processed in pixels
+   int64_t mMaxWidth {};
+   // This value is incremented on every lookup
+   uint64_t mCacheAccessIndex {};
+   // A multiplier used to control the cache size
+   int32_t mCacheSizeMultiplier { 4 };
+
+   template <typename CacheElementType>
```

Comment:
I get a clang build error:

```
GraphicsDataCache.h:213:23: error: declaration of 'CacheElementType' shadows template parameter
   template <typename CacheElementType>
```

Easily fixed by using another name (like just T)

Reply:
Build errors are fixed in later commits. Given that I have to build on 3 different systems I will postpone force pushes to the much later times